### PR TITLE
Add --content option

### PR DESCRIPTION
**Pull request by wclr**
This adds the ability to check if content of the files changes. There were [an issue](https://github.com/fgnass/node-dev/issues/200). I find it very useful and just necessary in my workflow because I run and watch compiled js files that may be accidentally touched by the compiler (often massively) and trigger restarts that are absolutely not needed.

I made it work like the `ignore` option, though I didn't add tests, but it is really straightforward.

Content is added to contentMap when the file is required (by reading file in async mode and calculating hash), then on change, it reads file content (in sync mode) and checks if the content hash is changed.

### CODE REVIEW

**Comment by bjornstar** [comment]
Could you please add some tests?

**Comment by wclr** [comment]
Ok, I will if you ok with the feature.

**Comment by bjornstar** [comment]
I'm not against the feature. I do have some issue with the synchronous file reading and calculating the md5 without using the stream interface.

**Comment by wclr** [comment]
@bjornstar ok, it is not difficult to make it async with streaming.

**Comment by wclr** [comment]
I made it async, and added tests that are passing, all seems to work fine for me. 

Could you look at tests that are not passing in ci? Maybe there is some logic error with calling correct calling t.end, but this test is passing on my machine.

Just for notice: 

1) It is not convenient, because there is a lot of excessive output.
2) Some tests are not passing on my machine.

File: test/spawn/index.js [hunk 1]
```diff
@@ -29,19 +30,47 @@ tap.test('should restart the server twice', t => {
   });
 });
 
-tap.test('should handle errors', t => {
```
Comment (bjornstar) [inline test/spawn/index.js:32]: Please leave the existing tests as they are.

Comment (wclr) [reply test/spawn/index.js:32]: My bad, just removed it by accident.

**Comment by bjornstar** [comment]
> 1. It is not convenient, because there is a lot of excessive output.
> 2. Some tests are not passing on my machine.

Thanks for your feedback, I've gone ahead and split the spawn test up into multiple files so that you can run the tests individually.

When you say the tests are not passing, what do you mean?

**Comment by wclr** [comment]
@bjornstar 

> Thanks for your feedback, I've gone ahead and split the spawn test up into multiple files so that you can run the tests individually.

I think I would be better to remove the output but make it optional with passing an argument. I run an individual test using tap cli: `npx tap --only test/spawn` (that will run only tests that are defined using `tap.only` - it is ok for development)

> When you say the tests are not passing, what do you mean?

They hang when I run `node test`

image

**Comment by bjornstar** [comment]
> I think it would be better to remove the output but make it optional with passing an argument. So that is the argument passed say npm test --output, then stdout/stderr from running node-dev (that is run by test/util/spawn) is shown, it would make running test output clean.

I'd rather have the output when we don't need it than not have the output when we do.

> I run individual tests using tap cli: npx tap --only --no-cov test/spawn (that will run only tests that are defined using tap.only - it is ok for development)

Looks like a lot of work and leaves us prone to committing `test.only`.

> They hang when I run `node test`
> 
> image

It appears to be the cluster test. I don't see any logs of the workers disconnecting which means that the test won't progress. I imagine there's something windows specific going on there. I'll see if I can reproduce it on my windows machine. I've been thinking about adding a windows environment to CI.

**Comment by wclr** [comment]
> Looks like a lot of work and leaves us prone to committing test.only.

Yes, that really happens, but it is kind of fixable using the commit hook, which would check it.

**Comment by wclr** [comment]
@bjornstar I've added the commit that prevents repeated restarts that may happen if multiple files are changed. Seems this bug was actual and before the current updates, not sure why there were no reports.

**Comment by bjornstar** [comment]
Could you please resolve the merge conflicts?

**Comment by wclr** [comment]
@bjornstar I've resolved, not sure what is wrong with tests on the CI.

On my machine, `./cluster` tests are handing.

**Comment by wclr** [comment]
@bjornstar 
Here is the issue I see with [added check isPaused](https://github.com/fgnass/node-dev/pull/254/files#diff-92bbac9a308cd5fcf9db165841f2d90ce981baddcb2b1e26cfff170929af3bd1R74).

This is certainly needed because it prevents duplicate restarts (if multiple files are modified). But the problem happens when a child does not exist after sending to it `SIGTERM`. In this case, `isPaused` stays `true`, and any restarts will not happen on change any more. What do you think about this scenario? How child exit failure (or non-exit) should be handled?

**Comment by wclr** [comment]
I may propose in case of a child didn't exist (after SIGTERM) send it **SIGKILL** after a timeout (for example defaulting 10 or 30 seconds).

**Comment by wclr** [comment]
@bjornstar So what would you say about adding SIGKILL timeout? It seems to work fine for me.

**Comment by bjornstar** [comment]
> @bjornstar So what would you say about adding SIGKILL timeout? It seems to work fine for me.

`isPaused` is for the `filewatcher` instance and could probably be read as `isRestarting`. The `filewatcher` usage should probably be changed to just use a new `filewatcher` instead of trying to reuse the existing one.

I've done some initial work based on your branch here -- https://github.com/bjornstar/node-dev/tree/content

It's almost there, I just haven't gotten around to finishing it.

**Comment by wclr** [comment]
@bjornstar Ok let's then see when you finish with this.

**Review by lastthyme — APPROVED** [review]

**Comment by bjornstar** [comment]
This seems to have fallen behind, feel free to pick it back up and re-open it if you're still interested in it.

**Pull request closed without merging**